data/tutorials/language/1ms_01_functors.md

```diff
@@ -8,183 +8,392 @@ category: "Module System"
-    val to_seq_from : elt -> t -> elt Seq.t
-    val to_seq : t -> elt Seq.t
-    val to_rev_seq : t -> elt Seq.t
-    val add_seq : elt Seq.t -> t -> t
-    val of_seq : elt Seq.t -> t
-  end
-```
-
-(the parentheses are necessary)
-
-## Defining Functors
-
-A functor with one argument can be defined like this:
-
-<!-- $MDX skip -->
-```ocaml
-module F (X : X_type) = struct
-  ...
-end
-```
-
-where `X` is the module that will be passed as argument, and `X_type` is its
-signature, which is mandatory.
-
-The signature of the returned module itself can be constrained, using this
-syntax:
-
-<!-- $MDX skip -->
-```ocaml
-module F (X : X_type) : Y_type =
-struct
-  ...
-end
-```
-
-or by specifying this in the `.mli` file:
-
-<!-- $MDX skip -->
-```ocaml
-module F (X : X_type) : Y_type
-```
-
-Overall, the syntax of functors is hard to grasp. The best may be to look at
-the source files
-[`set.ml`](https://github.com/ocaml/ocaml/blob/trunk/stdlib/set.ml) or
-[`map.ml`](https://github.com/ocaml/ocaml/blob/trunk/stdlib/map.ml) of the
-standard library.
+## Introduction
+
+Learning goals:
+- How to use a functor
+- How to write a functor
+- When to use a functor, some cases
+
+A functor is just a parametrised module.
+
+**Prerequisites**: Transitive closure leading to modules
+
+## Project Setup
+
+This tutorial uses the [Dune](https://dune.build) build tool. Make sure you have installed version 3.7 or later. We start by creating a fresh project. We need a folder named `funkt` with files `dune-project`, `dune`, `funkt.opam`, and `funkt.ml`. The latter two are created empty.
+```shell
+$ mkdir funkt; cd funkt
+
+$ touch funkt.opam funkt.ml
+```
+
+**`dune-project`**
+```lisp
+(lang dune 3.7)
+```
+
+**`dune`**
+```lisp
+(executable
+  (name funkt)
+  (public_name funkt)
+  (libraries str))
+```
+
+Check this works using the `dune exec funkt` command, it shouldn't do anything (the empty file is valid OCaml syntax) but it shouldn't fail either. The stanza `libraries str` will be used later.
+
+## Using an Existing Functor: `Set.Make`
+
+The standard library contains a [`Set`](/api/Set.html) module providing a data structure that allows operations like union and intersection. To use the provided type and its associated [functions](/api/Set.S.html), it's necessary to use the functor provided by `Set`. For reference only, here is a shortened version of the interface of `Set`:
+```ocaml
+module type OrderedType = sig
+  type t
+  val compare : t -> t -> int
+end
+
+module type S = sig
+  (* ... *)
+end
+
+module Make : functor (Ord : OrderedType) -> S
+```
+
+**Note**: Most set operation implementations must use a comparison function. Using `Stdlib.compare` would make it impossible to use a user-defined comparison algorithm. Passing the comparison function as a higher-order parameter (like in `Array.sort`, for instance) would add a lot of boilerplate. Providing set operations as a functor allows specifying the comparison function only once.
```

Comment:
"...a lot of boilerplate code." ??
boilerplate is usually an adjective, so would the modified noun be code? file?